### DISCUSSION TITLE
How can I open multi instance of an Mac app?

### DISCUSSION THREAD

**Discussion by ToTheBeginning** [pyinstaller/pyinstaller#6861] [PyInstaller] [answered]
I use pyinstaller to install an image viewer which is written by pyqt. After the installation, I change the default app to open an image file to the installed app. However, once an image is opened I can't open another image with the app. But I want the app function like the Preview app, i.e., once I double click a file, a new instance of the app is launched.

I found a [py2app solution](https://stackoverflow.com/questions/22898741/how-to-launch-multiple-instance-of-an-app-on-mac-osx), but it didn't work for me.

Can anyone help me with this? Thanks first.

**Comment by rokm** [comment, marked as answer]
App bundles don't support spawning multiple instances, but rather expect you to handle additional open request events once the app is started (and, for example, open new window within the running application instance).

If you're using Qt, you'll need to set up handler for [`QFileOpenEvent`](https://doc.qt.io/qt-5/qfileopenevent.html) - see also [this example](https://pyinstaller.org/en/stable/feature-notes.html#open-event-handling-in-a-qt-based-gui-application).

**Reply by ToTheBeginning** [reply]
Thank you so much. This is exactly what I need!

**Reply by Moosems** [reply]
How could I do this with tkinter?

**Reply by rokm** [reply]
> How could I do this with tkinter?

See [Tk docs](https://www.tcl.tk/man/tcl8.6/TkCmd/tk_mac.html) for general information on macOS-specific commands, as well as [this example from our docs](https://pyinstaller.org/en/stable/feature-notes.html#open-event-handling-in-a-tkinter-based-gui-application).

**Reply by Moosems** [reply]
Tk Docs doesn't have anything on this. Thanks for the idea though! It seems that tkinter in general also doesn't support multiple instances so that's *wonderful*

**Reply by bwoodsend** [reply]
You're kinda missing Rok's point. You don't create multiple instances of macOS .app applications no matter what GUI library you're using. Rather you have a single instance of the app which, on receiving an Apple Open File request, creates a new window (without spawning a new instance) to view/edit/process that new file. Whenever a window is closed, the .app instance remains running - even when all windows are closed - until the user selects right click->Quit on the application's icon in Dock.

**Reply by Moosems** [reply]
bwoodsend, I understood what he meant. What I was saying was that tkinter doesn't allow the user to bind a function to the DockMenu icon menubar
